src/par_iter/internal.rs

```diff
@@ -16,7 +16,12 @@ pub trait ProducerCallback<ITEM> {
 /// A producer which will produce a fixed number of items N. This is
 /// not queryable through the API; the consumer is expected to track
 /// it.
-pub trait Producer: IntoIterator + Send + Sized {
+pub trait Producer: Send + Sized
+{
+    type Item;
+    type IntoIter: Iterator<Item = Self::Item> + DoubleEndedIterator + ExactSizeIterator;
+    fn into_iter(self) -> Self::IntoIter;
```

Comment:
I don't love this change to an embedded `IntoIterator`, but I see how it's necessary to get the additional trait bounds to propagate implicitly.  It's yet another case of rust-lang/rust#20671, and actually kind of interesting to point out that associated types constraints do get elaborated this way, but they don't if you write:

```rust
pub trait Producer: IntoIterator + Send + Sized
    where <Self as IntoIterator>::IntoIter: DoubleEndedIterator + ExactSizeIterator
{...}
```

*sigh*